### DISCUSSION TITLE
How to seamlessly switch from data only channel to video/audio in multitrack conference without re-initializing webrtc adaptor

### DISCUSSION THREAD

**Discussion by tremzo** [ant-media/Ant-Media-Server#5367] [General]
When a user joins a room, they start with `webrtcadaptor.onlyDataChannel = true` (data channel only), Then they can start a video or audio call.
 
When sending the command `webrtcadaptor.onlyDataChannel = false; ` and then start a video or audio stream the Id of the A/V stream is changed to a new stream id (not the same id as when in datachannel only). 

It appears that when switching from `webrtcadaptor.onlyDataChannel = true` to `webrtcadaptor.onlyDataChannel = false`, the connection is disconnected and a new connection is established with a new stream ID of the A/V stream.
 
What is needed is to go from datachannel only mode to normal conference mode without closing the datachannel connection and reconnecting with `webrtcadaptor.onlyDataChannel = true`, basically dynamically switches from one mode to the other maintaining the connection and the same stream id.

It seems to me when starting a live stream with `onlyDataChannel=true` then no video/audio track is created, so trying to then turn on the video/audio when the data channel is already publishing, an error is triggered `error callback: "already_publishing"`.

So how is it possible to create a video track on the fly when already publishing the data channel?

**Comment by mekya** [comment]
Hi @timantmedia,

In order to have a seamless switch, I've created an issue here 
https://github.com/ant-media/Ant-Media-Server/issues/5392 and made some development. 
It will be in new release and next release will be in about 1 or 2 weeks. 

Fortunately, we've already merged the development and SNAPSHOT release is already available in npmjs for testing.  
[Check the live sample on codepen.io: Start WebRTC Publish without Opening Cam/Mic and Open Cam/Mic on the fly](https://codepen.io/mekya/pen/VwVEXNq)


Let me quick info for the developers:

1. Initialize the WebRTC adaptor with dummy audio and video as below
```js
var webRTCAdaptor = new WebRTCAdaptor({
  websocket_url: "wss://test.antmedia.io:5443/WebRTCAppEE/websocket",
  localVideoElement: document.getElementById("localVideo"),
  mediaConstraints: {
    video:"dummy",
    audio:"dummy"
  }
});
```

2. Publish the stream
```js
webRTCAdaptor.publish("your_stream_id");
```

3. Open camera and Mic on the fly
```js
webRTCAdaptor.openStream({video:true, audio:true}, "your_stream_id");
```

Please pay attention that second parameter is critical and it is the stream id of your current active broadcast. `webRTCAdaptor` uses that value to change the tracks of the peer connection on the fly.
